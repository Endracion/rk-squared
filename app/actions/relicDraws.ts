import { createAction } from 'typesafe-actions';

import { TimeT } from '../utils/timeUtils';

export interface RelicDrawBanner {
  id: number;
  imageUrl: string;
  openedAt: TimeT;
  closedAt: TimeT;
  sortOrder: number;

  canSelect: boolean;
  canPull: boolean;

  exchangeShopId?: number;

  group?: string;
  bannerRelics?: number[];
}

export interface RelicDrawGroup {
  groupName: string; // internal (not human-friendly) group name
  imageUrl: string;
  sortOrder: number;
}

export interface RelicDrawProbabilities {
  byRarity: {
    [rarity: number]: number;
  };
  byRelic: {
    [relicId: number]: number;
  };
}

export const setRelicDrawBanners = createAction(
  'SET_RELIC_DRAW_BANNERS',
  (banners: RelicDrawBanner[]) => ({
    type: 'SET_RELIC_DRAW_BANNERS',
    payload: banners,
  }),
);

export const setRelicDrawGroups = createAction(
  'SET_RELIC_DRAW_GROUPS',
  (groups: RelicDrawGroup[]) => ({
    type: 'SET_RELIC_DRAW_GROUPS',
    payload: groups,
  }),
);

export const setRelicDrawProbabilities = createAction(
  'SET_RELIC_DRAW_PROBABILITIES',
  (bannerId: number, probabilities: RelicDrawProbabilities) => ({
    type: 'SET_RELIC_DRAW_PROBABILITIES',
    payload: {
      bannerId,
      probabilities,
    },
  }),
);

export type RelicDrawAction = ReturnType<
  typeof setRelicDrawBanners | typeof setRelicDrawGroups | typeof setRelicDrawProbabilities
>;
