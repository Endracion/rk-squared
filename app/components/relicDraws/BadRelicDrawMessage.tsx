import * as React from 'react';

export const BadRelicDrawMessage = () => (
  <>
    <p>Relic draw details are not available, or this banner selection has expired.</p>
    <p>Please restart FFRK then go under Relic Draws.</p>
  </>
);
