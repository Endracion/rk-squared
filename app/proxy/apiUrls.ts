/**
 * @file
 * URLs for FFRK APIs
 */

const baseUrl = 'http://ffrk.denagames.com/dff/';

export const dungeons = (worldId: number) => `${baseUrl}world/dungeons?world_id=${worldId}`;
