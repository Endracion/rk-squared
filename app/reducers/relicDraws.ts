import { produce } from 'immer';
import { getType } from 'typesafe-actions';

import * as _ from 'lodash';

import { defaultOptions } from '../actions/options';
import {
  clearWantedRelics,
  ExchangeShopSelections,
  expireOldRelicDrawBanners,
  RelicDrawAction,
  RelicDrawBanner,
  RelicDrawGroup,
  RelicDrawProbabilities,
  setExchangeShopSelections,
  setRelicDrawBannersAndGroups,
  setRelicDrawProbabilities,
  wantRelic,
} from '../actions/relicDraws';
import { RealmRelicDrawMythrilCost } from '../data/probabilities';
import { arrayify } from '../utils/typeUtils';

export interface RelicDrawState {
  banners: {
    [bannerId: number]: RelicDrawBanner;
  };
  groups: {
    [group: string]: RelicDrawGroup;
  };
  probabilities: {
    // should be [bannerId: number], but string is easier for Lodash to work with
    [bannerId: string]: RelicDrawProbabilities;
  };
  selections: {
    [exchangeShopId: number]: ExchangeShopSelections;
  };
  want?: {
    [relicId: number]: boolean;
  };
}

const initialState: RelicDrawState = {
  banners: {},
  groups: {},
  probabilities: {},
  selections: {},
  want: {},
};

export function mergeFirstMythrilCost(
  prevBanner: RelicDrawBanner,
  newBanner: RelicDrawBanner,
): RelicDrawBanner | null {
  const prevCost = prevBanner.cost;
  const newCost = newBanner.cost;
  if (!prevCost || !newCost) {
    return null;
  }
  const firstMythrilCost = _.min(
    _.filter(
      [
        prevCost.mythrilCost,
        prevCost.firstMythrilCost,
        newCost.mythrilCost,
        newCost.firstMythrilCost,
      ],
      i => i != null,
    ),
  );
  if (firstMythrilCost == null || firstMythrilCost === newCost.mythrilCost) {
    return null;
  }
  return {
    ...newBanner,
    cost: {
      ...newBanner.cost,
      firstMythrilCost,
    },
  };
}

export function mergeBannersFirstMythrilCost(
  newBanners: {
    [bannerId: number]: RelicDrawBanner;
  },
  prevBanners: {
    [bannerId: number]: RelicDrawBanner;
  },
) {
  _.forEach(newBanners, (newBanner, id) => {
    if (prevBanners[+id]) {
      const update = mergeFirstMythrilCost(prevBanners[+id], newBanner);
      if (update) {
        newBanners[+id] = update;
      }
    }
  });
}

/**
 * Realm Relic Draws were refreshed before tracking of mythrilCost and
 * firstMythrilCost was implemented, so backfill our data structures with what
 * we know those to be.
 */
export function updateGroupFirstMythrilCosts(
  banners: { [bannerId: number]: RelicDrawBanner },
  group: string,
  firstMythrilCost: number,
) {
  _.forEach(banners, (banner, id) => {
    if (
      banner.group !== group ||
      !banner.cost ||
      !banner.cost.mythrilCost ||
      banner.cost.mythrilCost === firstMythrilCost
    ) {
      return;
    }
    banners[+id] = {
      ...banner,
      cost: {
        ...banner.cost,
        firstMythrilCost,
      },
    };
  });
}

export function relicDraws(
  state: RelicDrawState = initialState,
  action: RelicDrawAction,
): RelicDrawState {
  return produce(state, (draft: RelicDrawState) => {
    switch (action.type) {
      case getType(setRelicDrawBannersAndGroups): {
        const newBanners = _.keyBy(action.payload.banners, 'id');

        mergeBannersFirstMythrilCost(newBanners, state.banners);
        updateGroupFirstMythrilCosts(newBanners, 'group7', RealmRelicDrawMythrilCost);

        Object.assign(draft.banners, newBanners);
        Object.assign(draft.groups, _.keyBy(action.payload.groups, 'groupName'));

        return;
      }

      case getType(setRelicDrawProbabilities):
        draft.probabilities[action.payload.bannerId] = action.payload.probabilities;
        return;

      case getType(setExchangeShopSelections): {
        const { exchangeShopId, selections } = action.payload;
        draft.selections = draft.selections || {};
        draft.selections[exchangeShopId] = selections;
        return;
      }

      case getType(expireOldRelicDrawBanners): {
        const maxAgeInDays =
          action.payload.maxAgeInDays != null
            ? action.payload.maxAgeInDays
            : defaultOptions.maxOldRelicDrawBannerAgeInDays;

        const minClosedAt = action.payload.currentTime / 1000 - maxAgeInDays * 24 * 3600;

        draft.banners = _.omitBy(
          draft.banners,
          (value: RelicDrawBanner) => value.closedAt < minClosedAt,
        );

        const usedGroups = new Set<string>(_.map(draft.banners, i => i.group).filter(
          i => i != null,
        ) as string[]);
        draft.groups = _.pickBy(draft.groups, (value, key) => usedGroups.has(key));

        draft.probabilities = _.pickBy(
          draft.probabilities,
          (value, key) => draft.banners[+key] != null,
        );

        // This would be the logical place to also expire old exchange shop
        // selections, but we want to keep those around so that we can show
        // which selections are new.
        return;
      }

      case getType(wantRelic): {
        const relicIds = arrayify(action.payload.relicId);
        draft.want = draft.want || {};
        if (!action.payload.want) {
          for (const id of relicIds) {
            delete draft.want[id];
          }
        } else {
          for (const id of relicIds) {
            draft.want[id] = true;
          }
        }
        return;
      }

      case getType(clearWantedRelics): {
        draft.want = draft.want || {};
        for (const id of action.payload) {
          delete draft.want[id];
        }
        return;
      }
    }
  });
}
