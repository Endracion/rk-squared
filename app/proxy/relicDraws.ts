/**
 * @file
 * Support for tracking relic banners (gacha banners)
 */
import { Store } from 'redux';

import * as _ from 'lodash';

import {
  RelicDrawBanner,
  RelicDrawGroup,
  RelicDrawProbabilities,
  setRelicDrawBanners,
  setRelicDrawGroups,
  setRelicDrawProbabilities,
} from '../actions/relicDraws';
import { LangType } from '../api/apiUrls';
import * as gachaSchemas from '../api/schemas/gacha';
import { relativeUrl } from '../data/urls';
import { IState } from '../reducers';
import { logger } from '../utils/logger';
import { getRequestLang, Handler, HandlerRequest } from './common';

interface RelicDrawBannerResults {
  banners: RelicDrawBanner[];
  groups: {
    [groupName: string]: RelicDrawGroup;
  };
}

export function convertBanner(
  lang: LangType,
  gacha: gachaSchemas.GachaSeriesList,
  group?: string,
): RelicDrawBanner {
  return {
    id: gacha.series_id,
    openedAt: gacha.opened_at,
    closedAt: gacha.closed_at,
    sortOrder: gacha.priority,

    canPull: _.some(gacha.box_list, i =>
      _.some(
        i.entry_point_list,
        j => j.term_limit_num === 0 || j.purchased_count < j.term_limit_num,
      ),
    ),
    canSelect:
      gacha.total_executable_num > 0 &&
      gacha.user_exchange_shop_exchanged_num < gacha.total_executable_num,

    bannerRelics: _.map(gacha.banner_list, 'item_id').filter(i => i !== 0),

    exchangeShopId: +gacha.exchange_shop_id || undefined,
    imageUrl: relativeUrl(lang, gacha.line_up_image_path),
    group,
  };
}

export function convertRelicDrawBanners(
  lang: LangType,
  { gacha_group, series_list }: gachaSchemas.GachaShow,
): RelicDrawBannerResults {
  const result: RelicDrawBannerResults = {
    banners: [],
    groups: {},
  };
  const sortedSeries = _.keyBy(series_list, 'series_id');

  const processed = new Set<number>();

  function process(series: gachaSchemas.GachaSeriesList, group?: string) {
    if (!processed.has(series.series_id)) {
      result.banners.push(convertBanner(lang, series, group));
      processed.add(series.series_id);
    }
  }

  for (const group of gacha_group) {
    const thisGroup: RelicDrawGroup = {
      groupName: 'group' + group.id,
      imageUrl: relativeUrl(lang, group.line_up_image_path),
      sortOrder: group.priority,
    };
    result.groups[thisGroup.groupName] = thisGroup;
    group.content_series_ids
      .map(i => sortedSeries[i])
      .filter(i => i != null)
      .forEach(i => process(i, thisGroup.groupName));
  }

  const archiveSeries = series_list.filter(i => i.is_book_gacha);
  if (archiveSeries.length) {
    const thisGroup: RelicDrawGroup = {
      groupName: 'archive',
      imageUrl: relativeUrl(
        lang,
        '/dff/static/lang/image/gacha_series/lineup_image/book_mission_gacha.png',
      ),
      sortOrder: -100,
    };
    result.groups[thisGroup.groupName] = thisGroup;
    archiveSeries.forEach(i => process(i, thisGroup.groupName));
  }

  // These IDs can be extracted from app.js's REGULAR_SERIES_GACHA constants
  const realmRelicDrawId = 687;
  const seriesCount = 17;
  const realmRelicDraws = series_list.filter(
    i => i.series_id >= realmRelicDrawId && i.series_id < realmRelicDrawId + seriesCount,
  );
  if (realmRelicDraws.length) {
    const thisGroup: RelicDrawGroup = {
      groupName: 'realmRelicDraws',
      imageUrl: relativeUrl(
        lang,
        '/dff/static/lang/image/gacha_series/lineup_image/series_687.png',
      ),
      sortOrder: -99,
    };
    result.groups[thisGroup.groupName] = thisGroup;
    realmRelicDraws.forEach(i => process(i, thisGroup.groupName));
  }

  series_list.filter(i => i.priority !== 0).forEach(i => process(i));

  return result;
}

export function convertRelicDrawProbabilities(
  data: gachaSchemas.GachaProbability,
): RelicDrawProbabilities | null {
  const entryPointIds = _.keys(data).filter(i => i.match(/^\d+$/));
  if (entryPointIds.length === 0) {
    logger.error('Failed to find entry point ID for gacha/probability');
    return null;
  } else if (entryPointIds.length > 1) {
    logger.warn(`Unexpected entry point IDs for gacha/probability: got ${entryPointIds.length}`);
  }
  const entryPointId = +entryPointIds[0];
  const { prob_by_rarity, equipments } = data[entryPointId];

  return {
    byRarity: _.mapValues(prob_by_rarity, parseFloat),
    byRelic: _.fromPairs(
      equipments.filter(i => i.rarity >= 5).map(i => [i.id, parseFloat(i.probability)]),
    ),
  };
}

const gachaHandler: Handler = {
  'gacha/show'(data: gachaSchemas.GachaShow, store: Store<IState>, request: HandlerRequest) {
    const { banners, groups } = convertRelicDrawBanners(getRequestLang(request), data);
    store.dispatch(setRelicDrawBanners(banners));
    store.dispatch(setRelicDrawGroups(_.values(groups)));
  },

  'gacha/probability'(
    data: gachaSchemas.GachaProbability,
    store: Store<IState>,
    { query }: HandlerRequest,
  ) {
    if (!query || !query.series_id) {
      logger.error('Unrecognized gacha/probability query');
      return;
    }

    const probabilities = convertRelicDrawProbabilities(data);
    if (!probabilities) {
      return;
    }
    store.dispatch(setRelicDrawProbabilities(query.series_id, probabilities));
  },
};

export default gachaHandler;
