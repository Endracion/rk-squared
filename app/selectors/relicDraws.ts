import { createSelector } from 'reselect';

import * as _ from 'lodash';

import {
  ExchangeShopSelections,
  RelicDrawBanner,
  RelicDrawGroup,
  RelicDrawProbabilities,
} from '../actions/relicDraws';
import { enlir } from '../data/enlir';
import { IState } from '../reducers';
import { RelicDrawState } from '../reducers/relicDraws';
import { isClosed } from '../utils/timeUtils';
import { getOwnedLegendMateria, getOwnedSoulBreaks } from './characters';

export interface RelicDrawBannerDetails extends RelicDrawBanner {
  totalCount?: number;
  dupeCount?: number;
  selections?: ExchangeShopSelections;
}

export interface RelicDrawGroupDetails extends RelicDrawGroup {
  bannerCount: number;
  closedAt: number;
  canPull: boolean;
  canSelect: boolean;
  canPullOrSelectCount: number;
}

export type RelicDrawBannerOrGroup = RelicDrawBannerDetails | RelicDrawGroupDetails;

export function isGroup(item: RelicDrawBannerOrGroup): item is RelicDrawGroupDetails {
  return 'groupName' in item;
}

function getDupeCount(
  relicIds: number[],
  ownedSoulBreaks: Set<number> | undefined,
  ownedLegendMateria: Set<number> | undefined,
): number | undefined {
  if (!ownedSoulBreaks || !ownedLegendMateria) {
    return undefined;
  }

  return (
    _.filter(
      relicIds,
      i => enlir.relicSoulBreaks[i] && ownedSoulBreaks.has(enlir.relicSoulBreaks[i].id),
    ).length +
    _.filter(
      relicIds,
      i => enlir.relicLegendMateria[i] && ownedLegendMateria.has(enlir.relicLegendMateria[i].id),
    ).length
  );
}

export interface RelicDrawBannersAndGroups {
  // Indexed by group name - 'undefined' if no group
  [group: string]: RelicDrawBannerOrGroup[];
}

function getOneBannerDetails(
  banner: RelicDrawBanner,
  probabilities: RelicDrawProbabilities | undefined,
  allSelections: { [exchangeShopId: number]: ExchangeShopSelections },
  ownedSoulBreaks: Set<number> | undefined,
  ownedLegendMateria: Set<number> | undefined,
): RelicDrawBannerDetails {
  const selections =
    banner.exchangeShopId && allSelections ? allSelections[banner.exchangeShopId] : undefined;

  if (banner.bannerRelics && banner.bannerRelics.length !== 0) {
    return {
      ...banner,
      selections,
      totalCount: banner.bannerRelics.length,
      dupeCount: getDupeCount(banner.bannerRelics, ownedSoulBreaks, ownedLegendMateria),
    };
  } else if (probabilities) {
    const allRelics = _.keys(probabilities.byRelic).map(j => +j);
    return {
      ...banner,
      selections,
      totalCount: allRelics.length,
      dupeCount: getDupeCount(allRelics, ownedSoulBreaks, ownedLegendMateria),
    };
  } else if (selections) {
    return {
      ...banner,
      selections,
    };
  } else {
    return banner;
  }
}

export const getBannerDetails = createSelector<
  IState,
  RelicDrawState,
  Set<number> | undefined,
  Set<number> | undefined,
  { [bannerId: number]: RelicDrawBannerDetails }
>(
  (state: IState) => state.relicDraws,
  getOwnedSoulBreaks,
  getOwnedLegendMateria,
  ({ banners, probabilities, selections }, ownedSoulBreaks, ownedLegendMateria) => {
    return _.mapValues(banners, (banner: RelicDrawBanner) =>
      getOneBannerDetails(
        banner,
        probabilities[banner.id],
        selections,
        ownedSoulBreaks,
        ownedLegendMateria,
      ),
    );
  },
);

export const getBannersAndGroups = createSelector<
  IState,
  RelicDrawState,
  { [bannerId: number]: RelicDrawBannerDetails },
  RelicDrawBannersAndGroups
>(
  (state: IState) => state.relicDraws,
  getBannerDetails,
  ({ banners, groups }, bannerDetails) => {
    const result: { [group: string]: RelicDrawBannerOrGroup[] } = {};

    for (const group of [..._.keys(groups), undefined]) {
      const groupName = '' + group;

      result[groupName] = _.filter(bannerDetails, i => i.group === group);

      // If this is the root (undefined) group, then extend with all child groups.
      if (!group) {
        result[groupName].push(
          ..._.values(groups).map(g => {
            const groupBanners = _.filter(banners, i => i.group === g.groupName);
            return {
              ...g,
              bannerCount: groupBanners.length,
              closedAt: _.max(groupBanners.map(i => i.closedAt))!,
              canPull: _.some(groupBanners, i => i.canPull),
              canSelect: _.some(groupBanners, i => i.canSelect),
              canPullOrSelectCount: _.sumBy(groupBanners, i => +(i.canPull || i.canSelect)),
            };
          }),
        );
      }

      result[groupName] = _.sortBy(result[groupName], i => -i.sortOrder);
    }

    return result;
  },
);

export const getMissingBanners = createSelector<IState, RelicDrawState, number, number[]>(
  (state: IState) => state.relicDraws,
  (state: IState) => state.timeState.currentTime,
  ({ banners, probabilities, selections }, currentTime) => {
    // Start with all open banner IDs.
    const missing = new Set(
      _.keys(banners)
        .filter(i => !isClosed(banners[+i], currentTime))
        .map(i => +i),
    );

    // Remove banner IDs for which we have probabilities and don't need selections.
    const needsSelection = (bannerId: number) =>
      banners[bannerId] &&
      banners[bannerId].exchangeShopId != null &&
      !selections[banners[bannerId].exchangeShopId!];
    _.keys(probabilities)
      .filter(i => !needsSelection(+i))
      .forEach(i => missing.delete(+i));

    return Array.from(missing);
  },
);
