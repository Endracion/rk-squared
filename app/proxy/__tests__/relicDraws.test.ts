import { convertRelicDrawBanners, convertRelicDrawProbabilities } from '../relicDraws';

import * as _ from 'lodash';

import { LangType } from '../../api/apiUrls';

describe('gacha proxy handler', () => {
  describe('gacha/show', () => {
    it('converts relic draws with Realms on Parade', () => {
      const { data } = require('./data/gacha_show.json');
      const { banners, groups } = convertRelicDrawBanners(LangType.Gl, data);

      expect(_.values(banners).length).toEqual(45);

      const ff5BannerRelics = _.find(banners, i => i.id === 788)!.bannerRelics!;
      expect(ff5BannerRelics.length).toEqual(14);
      expect(ff5BannerRelics).toEqual([
        22050094,
        21008231,
        22053364,
        21001136,
        21010068,
        21008209,
        22051121,
        22055051,
        21006060,
        21004035,
        22056216,
        22053274,
        21009074,
        22056183,
      ]);

      expect(_.values(groups)).toEqual([
        {
          groupName: 'group4',
          imageUrl:
            'http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/gacha_series/lineup_image/series_happy_805.png',
          sortOrder: 1903081700,
        },
        {
          groupName: 'archive',
          imageUrl:
            'http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/gacha_series/lineup_image/book_mission_gacha.png',
          sortOrder: -100,
        },
        {
          groupName: 'realmRelicDraws',
          imageUrl:
            'http://ffrk.denagames.com/dff/static/lang/ww/compile/en/image/gacha_series/lineup_image/series_687.png',
          sortOrder: -99,
        },
      ]);
    });
  });

  describe('gacha/probability', () => {
    it('converts gacha probabilities', () => {
      const { data } = require('./data/gacha_probability.json');
      const probabilities = convertRelicDrawProbabilities(data);
      expect(probabilities).toBeTruthy();
      const { byRarity, byRelic } = probabilities!;
      expect(byRarity).toEqual({
        '3': 60.96,
        '4': 25.0,
        '5': 8.01999,
        '6': 6.01999,
      });
      expect(_.filter(byRelic, value => value >= 1).length).toEqual(14);
      expect(_.keys(byRelic).length).toEqual(42);
    });
  });
});
