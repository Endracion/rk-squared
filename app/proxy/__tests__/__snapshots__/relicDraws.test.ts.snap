// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gacha proxy handler exchange_shop/prize_list converts Dream Selection lists 1`] = `
Array [
  Object {
    "payload": Object {
      "exchangeShopId": "88",
      "selections": Array [
        Array [
          21002299,
        ],
        Array [
          22054079,
        ],
        Array [
          21002339,
        ],
        Array [
          21009065,
        ],
        Array [
          21004042,
        ],
        Array [
          22053257,
        ],
        Array [
          22055070,
        ],
        Array [
          21002243,
        ],
        Array [
          21010055,
        ],
        Array [
          21004049,
        ],
        Array [
          21006084,
        ],
        Array [
          21009082,
        ],
        Array [
          22055083,
        ],
        Array [
          21006092,
        ],
        Array [
          21007092,
        ],
        Array [
          21008156,
        ],
        Array [
          21003069,
        ],
        Array [
          21009104,
        ],
        Array [
          21007107,
        ],
        Array [
          21004047,
        ],
        Array [
          21001111,
        ],
        Array [
          21002209,
        ],
        Array [
          21009100,
        ],
        Array [
          21009110,
        ],
        Array [
          21008173,
        ],
        Array [
          22052047,
        ],
        Array [
          21002177,
        ],
        Array [
          22050057,
        ],
        Array [
          21006082,
        ],
        Array [
          21012022,
        ],
        Array [
          21010051,
        ],
        Array [
          21011025,
        ],
        Array [
          21008140,
        ],
        Array [
          21009095,
        ],
        Array [
          21003035,
        ],
        Array [
          21004034,
        ],
        Array [
          21002292,
        ],
        Array [
          21001130,
        ],
        Array [
          22053158,
        ],
        Array [
          21008189,
        ],
        Array [
          21005022,
        ],
        Array [
          21003047,
        ],
        Array [
          22056086,
        ],
        Array [
          21001039,
        ],
        Array [
          21001123,
        ],
        Array [
          21010039,
        ],
        Array [
          22053126,
        ],
        Array [
          21002248,
        ],
        Array [
          21008136,
        ],
        Array [
          21012023,
        ],
        Array [
          21002298,
        ],
        Array [
          21002297,
        ],
        Array [
          21002238,
        ],
        Array [
          21001109,
        ],
        Array [
          21001119,
        ],
        Array [
          21002337,
        ],
        Array [
          21011035,
        ],
        Array [
          21008172,
        ],
        Array [
          21015045,
        ],
        Array [
          21007113,
        ],
        Array [
          21003048,
        ],
        Array [
          21003055,
        ],
        Array [
          21013047,
        ],
        Array [
          21008087,
        ],
        Array [
          22053288,
        ],
        Array [
          21002233,
        ],
        Array [
          22056126,
        ],
        Array [
          21002090,
        ],
        Array [
          21032010,
        ],
        Array [
          21007074,
        ],
        Array [
          22056055,
        ],
        Array [
          21008113,
        ],
        Array [
          21008197,
        ],
        Array [
          22031008,
        ],
        Array [
          21013039,
        ],
        Array [
          21013060,
        ],
        Array [
          21011024,
        ],
        Array [
          21015058,
        ],
        Array [
          21002168,
        ],
        Array [
          22053262,
        ],
        Array [
          21003073,
        ],
        Array [
          21006095,
        ],
        Array [
          21008107,
        ],
        Array [
          21002400,
        ],
        Array [
          21002306,
        ],
        Array [
          21013067,
        ],
        Array [
          21007078,
        ],
        Array [
          21008123,
        ],
        Array [
          21015046,
        ],
        Array [
          21002212,
        ],
        Array [
          21002319,
        ],
        Array [
          21015047,
        ],
        Array [
          21013042,
        ],
        Array [
          21013068,
        ],
        Array [
          21002309,
        ],
        Array [
          21003070,
        ],
        Array [
          21008101,
        ],
        Array [
          21009112,
        ],
        Array [
          21002161,
        ],
        Array [
          21006102,
        ],
        Array [
          21006068,
        ],
        Array [
          21011008,
        ],
        Array [
          21007053,
        ],
        Array [
          21002265,
        ],
        Array [
          22053269,
        ],
        Array [
          21030009,
        ],
        Array [
          21008137,
        ],
        Array [
          21008147,
        ],
        Array [
          21030006,
        ],
        Array [
          21034006,
        ],
        Array [
          22056141,
        ],
        Array [
          21007046,
        ],
        Array [
          21007081,
        ],
        Array [
          21003078,
        ],
        Array [
          21003026,
        ],
        Array [
          22053277,
        ],
        Array [
          21030012,
        ],
        Array [
          21009106,
        ],
        Array [
          21002206,
        ],
        Array [
          21008190,
        ],
        Array [
          21009071,
        ],
        Array [
          21002189,
        ],
        Array [
          21001101,
        ],
        Array [
          21009103,
        ],
        Array [
          21003064,
        ],
        Array [
          21003065,
        ],
        Array [
          21015054,
        ],
        Array [
          21010052,
        ],
        Array [
          21010057,
        ],
        Array [
          21005018,
        ],
        Array [
          22050077,
        ],
        Array [
          21002207,
        ],
        Array [
          22053170,
        ],
        Array [
          21002279,
        ],
        Array [
          22053287,
        ],
        Array [
          21002074,
        ],
        Array [
          21002355,
        ],
        Array [
          21002102,
        ],
        Array [
          21007106,
        ],
        Array [
          21008171,
        ],
        Array [
          21009056,
        ],
        Array [
          21015041,
        ],
        Array [
          21013031,
        ],
        Array [
          21006076,
        ],
        Array [
          21010048,
        ],
        Array [
          21007122,
        ],
        Array [
          21002253,
        ],
        Array [
          21012025,
        ],
        Array [
          21009098,
        ],
        Array [
          21007105,
        ],
        Array [
          21014010,
        ],
        Array [
          21002270,
        ],
        Array [
          21015066,
        ],
        Array [
          21008174,
        ],
        Array [
          22050081,
        ],
        Array [
          21006108,
        ],
        Array [
          21002323,
        ],
        Array [
          21001134,
        ],
        Array [
          21015064,
        ],
        Array [
          21013070,
        ],
        Array [
          22056193,
        ],
        Array [
          21003042,
        ],
        Array [
          21002386,
        ],
        Array [
          21002303,
        ],
        Array [
          21002262,
        ],
        Array [
          21001113,
        ],
        Array [
          21009108,
        ],
        Array [
          21015061,
        ],
        Array [
          21002351,
        ],
        Array [
          21004040,
        ],
        Array [
          21008163,
        ],
        Array [
          21008164,
        ],
        Array [
          22053254,
        ],
        Array [
          21002330,
        ],
        Array [
          21009114,
        ],
        Array [
          21009120,
        ],
        Array [
          21013062,
        ],
        Array [
          21011031,
        ],
        Array [
          21006103,
        ],
        Array [
          21002343,
        ],
        Array [
          21001105,
        ],
        Array [
          21002341,
        ],
        Array [
          21015076,
        ],
        Array [
          21005032,
        ],
        Array [
          21012032,
        ],
        Array [
          21004056,
        ],
        Array [
          21002281,
        ],
        Array [
          21008153,
        ],
        Array [
          21014013,
        ],
        Array [
          22053163,
        ],
        Array [
          21014014,
        ],
      ],
    },
    "type": "SET_EXCHANGE_SHOP_SELECTIONS",
  },
]
`;
